### PR TITLE
HSEARCH-3074 Avoid depending on internal Hibernate ORM types

### PR DESCRIPTION
**Pull request by fax4ever**
https://hibernate.atlassian.net/browse/HSEARCH-3074

### CODE REVIEW

File: integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/massindexing/MassIndexingEmbeddedIdIT.java [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+	}
+
+	@Entity
+	@Table(name = "book")
+	@Indexed(index = Book.INDEX)
+	public static class Book {
+		public static final String INDEX = "Book";
+
+		@EmbeddedId
+		private BookId id;
```
Comment (yrodiere) [inline integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/massindexing/MassIndexingEmbeddedIdIT.java]: Could you maybe give another name to the ID property, such as "foobar"? I'm not sure using "id" in a query string will have Hibernate ORM auto-magically unrestand we're refering to the ID.

In particular, what if another property is named "id", but is not the entity id?

File: util/internal/integrationtest/orm/src/main/java/org/hibernate/search/util/impl/integrationtest/orm/SimpleSessionFactoryBuilder.java [hunk 2]
```diff
@@ -28,8 +28,8 @@
 	private final List<Consumer<MetadataSources>> metadataSourcesContributors = new ArrayList<>();
 	private final List<Consumer<SessionFactoryBuilder>> sessionFactoryBuilderContributors = new ArrayList<>();
 
-	public SimpleSessionFactoryBuilder setTcclLookupPrecedence(TcclLookupPrecedence tcclLookupPrecedence) {
-		return onBootstraServiceRegistryBuilder( builder -> builder.applyTcclLookupPrecedence( tcclLookupPrecedence ) );
+	public SimpleSessionFactoryBuilder setTcclLookupPrecedenceBefore() {
+		return onBootstraServiceRegistryBuilder( builder -> builder.applyClassLoader( new TcclLookupPrecedenceClassLoader() ) );
```
Comment (yrodiere) [inline util/internal/integrationtest/orm/src/main/java/org/hibernate/search/util/impl/integrationtest/orm/SimpleSessionFactoryBuilder.java]: I'm really not sure about this.

You're implementing your own classloader just to change the TCCL precedence in Hibernate ORM, which is a feature that is already exposed through SPIs through `org.hibernate.boot.registry.BootstrapServiceRegistryBuilder#applyTcclLookupPrecedence`. There's a fair amount of risk involved in deviating too much from the what our users' environment will be: our tests may end up not being representative.

Yes, `TcclLookupPrecedence` itself is marked as internal, but then it's exposed in `org.hibernate.boot.registry.BootstrapServiceRegistryBuilder` which is clearly an SPI. So in this case an implementation type is leaking through Hibernate ORM's SPI. The problem is in ORM, not in our code.

If you want to limit the usage of `TcclLookupPrecedence`, I'd advise to apply your changes up to this method, and give up here, calling `applyTcclLookupPrecedence( TcclLookupPrecedence.BEFORE )` directly.

If Hibernate ORM breaks this SPI, well, we can always fall back to your solution. It's not a big deal, as this is test code and a compatibility break will not affect our users.

File: integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/massindexing/MassIndexingEmbeddedIdIT.java (hunk 1 above)
Comment (fax4ever) [reply integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/massindexing/MassIndexingEmbeddedIdIT.java]: Good idea!

Comment (fax4ever) [reply integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/massindexing/MassIndexingEmbeddedIdIT.java]: You're right! If I change the name of the Hibernate ORM identifier, it works. But if I add a non-identifier field having `id` as a name, it doesn't!

File: util/internal/integrationtest/orm/src/main/java/org/hibernate/search/util/impl/integrationtest/orm/SimpleSessionFactoryBuilder.java (hunk 2 above)
Comment (fax4ever) [reply util/internal/integrationtest/orm/src/main/java/org/hibernate/search/util/impl/integrationtest/orm/SimpleSessionFactoryBuilder.java]: Is a SPI method that has an internal parameter a little strange? Anyway let's limit as much as possible the internal exposition.

Comment (fax4ever) [reply util/internal/integrationtest/orm/src/main/java/org/hibernate/search/util/impl/integrationtest/orm/SimpleSessionFactoryBuilder.java]: As the comment says... Actually it was my first intention.

**Comment by fax4ever** [comment]
@yrodiere thanks for the review. Anything else? Thanks

File: util/internal/integrationtest/orm/src/main/java/org/hibernate/search/util/impl/integrationtest/orm/SimpleSessionFactoryBuilder.java (hunk 2 above)
Comment (yrodiere) [reply util/internal/integrationtest/orm/src/main/java/org/hibernate/search/util/impl/integrationtest/orm/SimpleSessionFactoryBuilder.java]: > Is a SPI method that has an internal parameter a little strange

Yes I totally agree it's strange. But for now we'll have to deal with it.

**Comment by yrodiere** [comment]
@fax4ever Looks good! Maybe you could just add forbiddenapis rules to forbid using classes in packages `org.hibernate.internal.**` and `org.hibernate.**.internal.**`?
Just launch the build after your added the rules, to check it fails, then add the necessary exclusions in `pom.xml` (around line 1411).

**Comment by fax4ever** [comment]
@yrodiere thanks! I'll added the commit with the rule for the forbiddenapis plugin. I'll wait the CI, then I'll merge it.

**Comment by fax4ever** [comment]
Merged thanks!

**Pull request merged**